nilearn/surface/surface.py

```diff
@@ -1975,3 +1976,243 @@ def new_img_like(ref_img, data):
+    -------
+    matrix : scipy.sparse.csr_matrix
+        A sparse matrix representing the edge relationships in `surface`.
+
+    """
+    # This is a bit of a hack to quickly find a unique set of all edges.
+    if values not in {"len", "invlen", "ones"}:
+        raise ValueError(f"unrecognized values argument: {values}")
+
+    n = mesh.coordinates.shape[0]
+
+    edges = np.vstack(
+        [
+            mesh.faces[:, [0, 1]],
+            mesh.faces[:, [0, 2]],
+            mesh.faces[:, [1, 2]],
+        ]
+    )
+    edges = edges.astype(np.int64)
+    bigcol = edges[:, 0] > edges[:, 1]
+    lilcol = ~bigcol
+    edges = np.concatenate(
+        [
+            edges[bigcol, 0] + edges[bigcol, 1] * n,
+            edges[lilcol, 1] + edges[lilcol, 0] * n,
+        ]
+    )
+    edges = np.unique(edges)
+
+    (u, v) = (edges // n, edges % n)
+
+    # Calculate distances between pairs.
+    # We use this as a weighting to make sure that
+    # smoothing takes into account the distance between each vertex neighbor
+    if values in ("len", "invlen"):
+        coords = mesh.coordinates
+        edge_lens = np.sqrt(np.sum((coords[u, :] - coords[v, :]) ** 2, axis=1))
+        if dtype is None:
+            dtype = edge_lens.dtype
+        else:
+            edge_lens = edge_lens.astype(dtype)
+        if values == "invlen":
+            edge_lens = 1 / edge_lens
+    elif values == "ones":
+        if dtype is None:
+            edge_lens = np.ones_like(edges)
+        else:
+            edge_lens = np.ones(edges.shape, dtype=dtype)
+
+    # We can now make a sparse matrix.
+    ee = np.concatenate([edge_lens, edge_lens])
+    uv = np.concatenate([u, v])
+    vu = np.concatenate([v, u])
+
+    return csr_matrix((ee, (uv, vu)), shape=(n, n))
+
+
+def smooth_img(
+    imgs,
+    iterations=1,
+    distance_weights=False,
+    vertex_weights=None,
+    center_surround_knob=0,
+    match="sum",
+):
+    """Smooth values along the surface.
+
+    Parameters
+    ----------
+    imgs : SurfaceImage
+        The surface whose is to be smoothed.
+        In the case of 2D data, each sample is smoothed independently.
+
+    iterations : :obj:`int`, default = 1
+        The number of times to repeat the smoothing operation
+        (it must be a positive value).
+        Defaults to 1
+
+    distance_weights : :obj:`bool`, default = False
+        Whether to add distance-based weighting to the smoothing.
+        With such weights, the value calculated for each vertex
+        at each iteration is the weighted sum of neighboring vertices
+        where the weight on each neighbor is the inverse
+        of the distances to it.
+
+    vertex_weights : SurfaceImage or None, default = None
+        A SurfaceImage whose data are vector of weights, one per vertex.
+        These weights are normalized and
+        applied to the smoothing
+        after the application of center-surround weights.
+
+    center_surround_knob : :obj:`float`, default = 0
```

Comment:
I would not allow this to be tweaked: I would only average a given vertex with its nearest neighbour